File: archive/scripts/dl012_archive.py
"""DL-012 archival: Insert question_state: "Archived" for MISSING clone items.
Safe rebuild — parses all before writing. Per BACKUP_PROTOCOL.md Script Safety Rule.
"""
import re, sys, os

def archive_missing_clones(filepath, prefix):
    """Find Section E items without question_state and insert 'Archived'."""
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    # Pattern: find QuestionID for Section E items
    qid_pattern = re.compile(r'("QuestionID":\s*"' + re.escape(prefix) + r'-\d{3}")')
    state_pattern = re.compile(r'"question_state":\s*"([^"]*)"')

    matches = list(qid_pattern.finditer(content))
    print(f"  Found {len(matches)} QuestionIDs matching {prefix}-")

    insertions = []
    already_archived = 0
    certified_hits = 0

    for i, match in enumerate(matches):
        qid_m = re.search(r'P1-[A-Z]+-\d+', match.group(0))
        if not qid_m:
            continue
        qid = qid_m.group(0)
        idx = match.start()

        # Look 800 chars ahead for question_state
        next_boundary = content.find('"QuestionID"', idx + len(match.group()))
        if next_boundary == -1:
            window = content[idx:idx + 1200]
        else:
            window = content[idx:min(idx + 1200, next_boundary)]

        # Check if question_state exists
        sm = state_pattern.search(window)
        if sm:
            state_value = sm.group(1)
            if state_value == 'Certified':
                certified_hits += 1
                print(f"  *** CRITICAL: {qid} is Certified! HALT ***")
            elif state_value == 'Archived':
                already_archived += 1
            # else: has some other state (Unprocessed, etc.) — skip
            continue

        # MISSING: need to insert question_state
        # Find the end of the QuestionID line
        qid_line_end = content.index('\n', idx) + 1
        # Insert right after
        indent = '    '
        insertion = f'{indent}"question_state": "Archived",\n'
        insertions.append((qid_line_end, insertion, qid))

    print(f"  Already Archived: {already_archived}")
    print(f"  To archive: {len(insertions)}")
    print(f"  Certified hits: {certified_hits}")

    if certified_hits > 0:
        print("*** HALT: Certified items found. No writes performed. ***")
        return False, 0

    if len(insertions) == 0:
        print("  Nothing to do.")
        return True, 0

    # Apply insertions in reverse order so indices stay valid
    insertions.sort(key=lambda x: x[0], reverse=True)

    result = list(content)
    for pos, ins_text, qid in insertions:
        for ch in ins_text:
            result.insert(pos, ch)

    new_content = ''.join(result)

    # Verify the result
    verify_missing = len(re.findall(r'"QuestionID":\s*"' + re.escape(prefix) + r'-\d{3}"[\s\S]{0,800}?(?![\s\S]*?"question_state")', new_content))
    if verify_missing > 0:
        print(f"  *** VERIFY FAILED: {verify_missing} items still MISSING after insert ***")
        return False, 0

    print(f"  Write verified: 0 items still MISSING for {prefix}")

    with open(filepath, 'w', encoding='utf-8', newline='\n') as f:
        f.write(new_content)

    print(f"  Written: {filepath}")
    return True, len(insertions)


if __name__ == '__main__':
    base = r'C:\Users\User\OneDrive\Desktop\CMA_Part_1_2026'

    print("=== DL-012 Archival Script ===")
    print()

    # Pack C
    print("--- Pack C ---")
    ok_c, count_c = archive_missing_clones(
        os.path.join(base, 'pack_c_corrected.js'), 'P1-EC')

    # Pack D
    print("--- Pack D ---")
    ok_d, count_d = archive_missing_clones(
        os.path.join(base, 'pack_d_corrected.js'), 'P1-ED')

    print()
    total = count_c + count_d
    print(f"=== DONE: {total} items archived ===")
    print(f"Pack C: {count_c}, Pack D: {count_d}")
    sys.exit(0 if (ok_c and ok_d) else 1)

File: archive/scripts/dl012_archive.js
/**
 * DL-012 Archival: Insert question_state: "Archived" for MISSING Section E clone items.
 * Safe rebuild — parses all before writing. Per BACKUP_PROTOCOL.md Script Safety Rule.
 */
const fs = require('fs');
const path = require('path');

const BASE = 'C:\\Users\\User\\OneDrive\\Desktop\\CMA_Part_1_2026';

function archiveMissing(filepath, prefix) {
    const orig = fs.readFileSync(filepath, 'utf-8');

    // Find all QuestionID positions for this prefix
    const qidRe = new RegExp('"QuestionID":\\s*"' + prefix + '-\\d{3}"', 'g');
    const items = [];
    let m;
    while ((m = qidRe.exec(orig)) !== null) {
        const idx = m.index;
        const qid = m[0].match(/P1-[A-Z]+-\d+/)[0];

        // Scan next 1000 chars for question_state
        const nextQid = orig.indexOf('"QuestionID"', idx + m[0].length);
        const windowEnd = nextQid === -1 ? idx + 1000 : Math.min(idx + 1000, nextQid);
        const window = orig.substring(idx, windowEnd);
        const stateMatch = window.match(/"question_state":\s*"([^"]*)"/);

        items.push({
            idx,
            qid,
            hasState: !!stateMatch,
            stateValue: stateMatch ? stateMatch[1] : null,
            lineEnd: orig.indexOf('\n', idx)
        });
    }

    const total = items.length;
    const withState = items.filter(i => i.hasState).length;
    const archived = items.filter(i => i.stateValue === 'Archived').length;
    const certified = items.filter(i => i.stateValue === 'Certified');
    const missing = items.filter(i => !i.hasState);

    console.log(`  Found ${total} QuestionIDs for ${prefix}`);
    console.log(`  With question_state: ${withState}`);
    console.log(`  Already Archived: ${archived}`);
    console.log(`  To archive: ${missing.length}`);

    if (certified.length > 0) {
        certified.forEach(c => console.log(`  *** CRITICAL: ${c.qid} is Certified! ***`));
        console.log('*** HALT: Certified items found. No writes performed. ***');
        return false;
    }

    if (missing.length === 0) {
        console.log('  Nothing to do.');
        return true;
    }

    // Build new content with insertions (reverse order)
    const insertions = missing.map(item => ({
        pos: item.lineEnd + 1,
        text: '    "question_state": "Archived",\n',
        qid: item.qid
    }));

    insertions.sort((a, b) => b.pos - a.pos);

    let result = orig;
    for (const ins of insertions) {
        result = result.substring(0, ins.pos) + ins.text + result.substring(ins.pos);
    }

    // Verify: re-scan the same prefix
    const verifyQidRe = new RegExp('"QuestionID":\\s*"' + prefix + '-\\d{3}"', 'g');
    let stillMissing = 0;
    while ((m = verifyQidRe.exec(result)) !== null) {
        const idx = m.index;
        const nextQid2 = result.indexOf('"QuestionID"', idx + m[0].length);
        const wEnd = nextQid2 === -1 ? idx + 1000 : Math.min(idx + 1000, nextQid2);
        const win = result.substring(idx, wEnd);
        if (!win.match(/"question_state"/)) {
            stillMissing++;
        }
    }

    if (stillMissing > 0) {
        console.log(`  *** VERIFY FAILED: ${stillMissing} items still MISSING ***`);
        return false;
    }

    fs.writeFileSync(filepath, result, 'utf-8');
    console.log(`  Written: ${filepath}`);
    return true;
}

console.log('=== DL-012 Archival Script ===\n');

let okC = archiveMissing(path.join(BASE, 'pack_c_corrected.js'), 'P1-EC');
console.log('');
let okD = archiveMissing(path.join(BASE, 'pack_d_corrected.js'), 'P1-ED');

console.log('');
if (okC && okD) {
    console.log('=== DONE: All items archived successfully ===');
    process.exit(0);
} else {
    console.log('*** HALT: Errors detected ***');
    process.exit(1);
}
